### PR TITLE
Fix : Path tracker causes app to crash #362 (made changes after reviewed)

### PR DESCRIPTION
**Pull request by reverie-ss**
- [x] Apply the `MifosStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

Fix : Path tracker causes app to crash #362

Problem : App used to crash on clicking "Start Tracking " button in PathTrackingActivity
Reason : Location permission and request was not handled
Solution : Added run-time location permission request in PatchTrackingActivity

### CODE REVIEW

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java [hunk 1]
```diff
@@ -90,4 +118,51 @@ public void onServiceDisconnected(ComponentName componentName) {
...
+        }
+
+        try {
+            networkEnabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+        } catch (Exception ex) {
+        }
+
+        if (!gpsEnabled && !networkEnabled) {
+            // notify user
+            final AlertDialog.Builder builder = new AlertDialog.Builder(this);
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java]: Please use https://github.com/openMF/android-client/blob/master/mifosng-android/src/main/java/com/mifos/mifosxdroid/core/MaterialDialog.java for showing material dialog.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java [hunk 2]
```diff
@@ -40,19 +48,38 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
+            CheckSelfPermissionAndRequest
+                    .requestPermission(this,
+                            Manifest.permission.ACCESS_FINE_LOCATION,
+                            100,
+                            "Location Permission is necessary for this feature",
+                            "Location Permission is necessary for this feature",
+                            "Permission Denied");
+        } else {
+
+            start.setOnClickListener(new View.OnClickListener() {
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java]: we are using ButterKnife So we do not need to implement ```View.OnClickListener()```
Use ```onClick``` annotation instead.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java [hunk 3]
```diff
@@ -90,4 +118,51 @@ public void onServiceDisconnected(ComponentName componentName) {
...
+
+        try {
+            networkEnabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+        } catch (Exception ex) {
+        }
+
+        if (!gpsEnabled && !networkEnabled) {
+            // notify user
+            final AlertDialog.Builder builder = new AlertDialog.Builder(this);
+            builder.setMessage("Location seems to be disabled, do you want to enable it?")
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java]: Do not user hard coded strings.

**Comment by therajanmaurya** [comment]
@reverie-ss Please make above changes :)

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java (hunk 2 above)
Comment (reverie-ss) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java]: `start.setOnClickListener(new View.OnClickListener() {})`

It's there in the master code! I haven't added!

(https://github.com/openMF/android-client/blob/master/mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java#L43)

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java [hunk 4]
```diff
@@ -40,19 +48,38 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
-        });
-
-        stop.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View view) {
-                service.stopLogging();
-            }
-        });
+
+        if (!CheckSelfPermissionAndRequest
```
Comment (reverie-ss) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java]: Followed the implementation!
And I have added `onRequestPermissionsResult`

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java [hunk 5]
```diff
@@ -41,27 +50,33 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
         setContentView(R.layout.activity_path_tracker);
         ButterKnife.bind(this);
         start.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View view) {
-                service.startLogging();
-            }
-        });
+                @Override
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java]: Remove this View.OnClickListener implementation and use ButterKnife annotation.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java [hunk 6]
```diff
@@ -41,27 +50,33 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
+                }
+            });
 
         stop.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View view) {
-                service.stopLogging();
-            }
-        });
+                @Override
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java]: same as above

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java [hunk 7]
```diff
@@ -41,27 +50,33 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
+                public void onClick(View view) {
+                    service.stopLogging();
+                }
+            });
+
+
         showBackButton();
     }
 
+
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java]: do not add the blank extra line, remove this line.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java [hunk 8]
```diff
@@ -41,27 +50,33 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
     @Override
     protected void onStart() {
         super.onStart();
         Intent intent = new Intent(this, LocationService.class);
         startService(intent);
         bindService(intent, this, Context.BIND_AUTO_CREATE);
+
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/PathTrackingActivity.java]: same as above

**Comment by therajanmaurya** [comment]
This PR needs some changes please look comments.
and Before sending final PR please squash all your commits into one.

**Comment by reverie-ss** [comment]
-Fixed the issues in the comments
-Squashed all the commits into one

**I have made a new final PR request #418 . This PR can be closed.**
Thank you for your time.

**Pull request closed without merging**